### DISCUSSION TITLE
Adding support for a production mode.

### DISCUSSION THREAD

**Discussion by yopknopixx** [micropython/micropython#9386] [Using MicroPython]
I recently worked on a project with MicroPython where the performance was critical. I could get it working satisfactorily using various optimizations from [https://docs.micropython.org/en/latest/develop/optimizations.html](url) and the native function wrapper.
I wondered if there is a better way to deploy a finished project onto a board, by compiling the project code and only the necessary MicroPython code into a firmware file with obfuscation, etc.

I have a few queries regarding this:
 - Will this enable some performance enhancements?
 - Will it be more space and memory efficient?
 - Is there a need for this?

 I would be willing to work on this myself with guidance from the community.

**Comment by robert-hh** [comment]
Freezing code into the firmware may affect the performance in both ways:
- importing code is faster (allmost immediate), since if has not to be compiled and loaded into RAM.
- the RAM size available for the heap is larger, since the code stays in flash. That reduces the risk of memory errors due to heap fragmentation and can reduce the time used for garbage collection.
- for MCU with serially attached Flash, it may reduce the code execution speed, since the memory range needed to be cached is larger.

Given the various aspects, code importing is definitely faster. Everything else has to be tested with your code in a real usage scenario.

**Reply by yopknopixx** [reply]
What I'm thinking of is a feature that will compile everything to machine code including the user code and only the used modules from the micropython firmware to create a firmware file that can be used in a production setting. I can imagine it reducing the flash and RAM footprint, but will it also have performance improvements (Ideally comparable to CPP/Arduino)?
If so, I think it will be a great feature addition to the code base.

**Reply by robert-hh** [reply]
> a feature that will compile everything to machine code including the user code

How do you want to achieve compiling the user code into machine code. Do you think of using frozen bytecode? That would include code with the @micropython.native decorator. Removing not needed modules from the firmware reduces the flash footprint, but native code can be larger than bytecode, albeit faster.

**Reply by yopknopixx** [reply]
> Removing not needed modules from the firmware reduces the flash footprint, but native code can be larger than bytecode, albeit faster.

Agreed.

I am currently thinking of implementing this using the native code, but if there is a better way, I'm open to it.
